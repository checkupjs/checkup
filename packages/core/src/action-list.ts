import { ActionConfig, TaskConfig, ConfigValue } from './types/config';
import { Action } from './types/tasks';
import { parseConfigTuple } from './config';

/**
 * @class ActionList
 *
 * Represents a collection of actions related to a task
 */
export default class ActionList {
  _defaultActions: Action[];
  _actionConfig: ActionConfig;

  constructor(defaultActions: Action[], config: TaskConfig | undefined) {
    this._defaultActions = defaultActions;
    this._actionConfig = config?.actions || {};
  }

  get actionMessages() {
    let actionMessages: string[] = [];
    this.enabledActions.forEach((action) => {
      if (action.enabled) {
        actionMessages.push(action.message);
      }
    });

    return actionMessages;
  }

  get isActionable() {
    return this.actionMessages.length > 0;
  }

  get enabledActions() {
    return this._defaultActions
      .map((action: Action) => {
        if (this._actionConfig[action.name]) {
          let [enabled, configuredAction] = this._applyConfigForAction(
            action,
            this._actionConfig[action.name]
          );
          return enabled ? configuredAction : undefined;
        }

        if (action.enabled) {
          return action;
        }
      })
      .filter(Boolean) as Action[];
  }

  private _applyConfigForAction(
    action: Action,
    actionConfig: ConfigValue<{ threshold: number }>
  ): [boolean, Action] {
    let [enabled, value] = parseConfigTuple<{ threshold: number }>(actionConfig);

    if (typeof value.threshold === 'number') {
      action.threshold = value.threshold;
    }

    return [enabled, action];
  }
}
