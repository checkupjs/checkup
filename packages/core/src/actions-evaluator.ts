import { TaskConfig, ActionConfig } from './types/config';
import { parseConfigTuple } from './config';
import { Action2 } from './types/tasks';

export default class ActionsEvaluator {
  private actions: Action2[] = [];

  add(action: Action2) {
    this.actions.push(action);
  }

  evaluate(config: TaskConfig): Action2[] {
    let actionConfig: ActionConfig = config.actions ?? {};

    return this.actions.filter((action: Action2) => {
      let [enabled, value] = parseConfigTuple<{ threshold: number }>(actionConfig[action.name]);
      let threshold =
        value && typeof value.threshold === 'number' ? value.threshold : action.defaultThreshold;

      return enabled && action.input >= threshold;
    });
  }
}
