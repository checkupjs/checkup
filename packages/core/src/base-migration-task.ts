import BaseTask from './base-task';
import { TaskContext, TaskResultOptions } from './types/tasks';

const merge = require('lodash.merge');

export type FeatureId = string;
export type Feature = {
  featureName: string;
  message: string;
  helpUri: string;
};

/**
 * Creates a new instance of a BaseMigrationTask.
 *
 * @param migrationName The short name of the migration. Used to format result messages.
 * @param pluginName The name of the plugin this task is included in.
 * @param context The runtime task context passed to the Task.
 */
export default abstract class BaseMigrationTask extends BaseTask {
  protected features: Map<FeatureId, Feature>;

  /**
   * A list of feature names that represent this migration.
   *
   * @readonly
   * @memberof BaseMigrationTask
   */
  get featureNames() {
    return [
      ...new Set(
        [...this.features.values()].map((ruleMetadata: Feature) => ruleMetadata.featureName)
      ),
    ];
  }

  /**
   * Creates a new instance of a migration Task.
   *
   * @param migrationName The short name of the migration. Used to format result messages.
   * @param pluginName The name of the plugin this task is included in.
   * @param context The runtime task context passed to the Task.
   */
  constructor(public migrationName: string, pluginName: string, context: TaskContext) {
    super(pluginName, context);

    this.features = new Map<FeatureId, Feature>();
  }

  /**
   * Adds componennt data to the rule metadata.
   */
  protected addRuleComponentMetadata() {
    this.addRule({
      properties: {
        component: {
          name: 'migration',
          options: {
            sortBy: 'value',
            sortDirection: 'desc',
          },
        },
      },
    });
  }

  /**
   * Adds a feature definition to the Checkup rule metadata. Used for correctly associating features and results.
   *
   * @param featureId - The ID of the feature, such as the lint rule ID corresponding to the feature
   * @param feature - An object representing the feature's details
   * @param feature.feature - The name of the feature
   * @param feature.message - The user-friendly message
   * @param feature.helpUri - A URL to provide help documentation about the feature
   */
  protected addFeature(featureId: FeatureId, feature: Feature) {
    this.features.set(featureId, feature);

    this.addRuleProperties({
      features: this.featureNames,
    });
  }

  /**
   * Adds a feature-specific result object to the Checkup output. The result includes the feature
   * metadata that this result is associated with in the migration.
   *
   * @param feature - An object representing the feature's details
   * @param feature.feature - The name of the feature
   * @param feature.message - The user-friendly message
   * @param feature.helpUri - A URL to provide help documentation about the feature
   * @param options Additional options to pass to the result
   * @param options.location Specifies a location where the result occurred
   * @param options.properties A property bag named properties, which stores additional values on the result
   */
  protected addFeatureResult(feature: Feature, options?: TaskResultOptions) {
    this.addResult(
      `${this.migrationName} | ${feature.featureName} : ${feature.message}. More info: ${feature.helpUri}`,
      'review',
      'warning',
      merge({}, options, {
        properties: {
          migration: {
            name: this.migrationName,
            featureName: feature.featureName,
          },
        },
      })
    );
  }
}
