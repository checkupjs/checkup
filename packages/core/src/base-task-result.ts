import { Category, TaskMetaData } from './types';

const CATEGORY_SORT_MAP = {
  [Category.Meta]: 4,
  [Category.Insights]: 3,
  [Category.Migrations]: 2,
  [Category.Recommendations]: 1,
};

export default abstract class BaseTaskResult {
  meta: TaskMetaData;

  constructor(meta: TaskMetaData) {
    this.meta = meta;
  }

  compareTo(other: BaseTaskResult) {
    let { category } = other.meta.taskClassification;
    let thisCategory = CATEGORY_SORT_MAP[this.meta.taskClassification.category];
    let otherCategory = CATEGORY_SORT_MAP[category];

    if (thisCategory > otherCategory) {
      return 1;
    } else if (thisCategory < otherCategory) {
      return -1;
    }
    return 0;
  }
}
