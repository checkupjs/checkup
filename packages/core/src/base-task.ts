import { Task, TaskResult } from './types';

export default abstract class BaseTask implements Task {
  static taskName: string;
  static friendlyTaskName: string;

  args: any;

  constructor(args: any) {
    this.args = args;
  }

  abstract run(): Promise<TaskResult>;
}
