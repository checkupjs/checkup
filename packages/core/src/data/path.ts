export function trimCwd(path: string, cwd: string) {
  return path.replace(`${cwd}/`, '');
}

export function trimAllCwd(paths: string[], cwd: string) {
  return paths.map((path) => trimCwd(path, cwd));
}
