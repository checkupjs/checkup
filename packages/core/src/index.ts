export { default as BaseTask } from './base-task';
export { default as FileSearcherTask } from './file-searcher-task';
export { default as FileSearcher } from './searchers/file-searcher';
export { loadPlugins } from './loaders/plugin-loader';
export { getPackageJson } from './utils/get-package-json';
export { default as CheckupConfigService } from './configuration/checkup-config-service';
export { default as CosmiconfigLoaderFactory } from './configuration/loaders/cosmiconfig-loader-factory';
export { ui } from './utils/ui';
export { toPairs } from './utils/data-transformers';

export * from './types';
