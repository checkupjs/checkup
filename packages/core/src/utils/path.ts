import { isAbsolute, join, relative } from 'path';

export function toAbsolute(baseDir: string, filePath: string) {
  let absolutePath = filePath;

  if (!isAbsolute(filePath)) {
    absolutePath = join(baseDir, filePath);
  }

  return absolutePath;
}

export function toRelative(baseDir: string, filePath: string) {
  let relativePath = filePath;

  if (isAbsolute(relativePath)) {
    relativePath = relative(baseDir, filePath);
  }

  return relativePath;
}
