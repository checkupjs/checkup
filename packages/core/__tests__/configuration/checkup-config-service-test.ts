import { CheckupConfig, CheckupConfigFormat, CheckupConfigService } from '../../src';
import CheckupFixturifyProject from '@checkup/test-helpers/lib/checkup-fixturify-project';
import * as path from 'path';
import * as fs from 'fs';

describe('checkup-config-service', () => {
  let defaultConfig: CheckupConfig;

  beforeEach(() => {
    defaultConfig = {
      plugins: [],
      tasks: {},
    };
  });

  it('should load a config via a loader', async () => {
    const configService = await CheckupConfigService.load(async () => ({
      filepath: '.',
      config: defaultConfig,
      format: CheckupConfigFormat.JSON,
    }));

    expect(configService.get()).toStrictEqual(defaultConfig);
  });

  it('should throw an error if an invalid config is present and get is called', async () => {
    const configService = await CheckupConfigService.load(async () => ({
      filepath: '.',
      config: ({
        plugins: {},
        tasks: {
          someTask: 'foo',
        },
      } as unknown) as CheckupConfig,
      format: CheckupConfigFormat.JSON,
    }));

    expect(() => configService.get()).toThrowErrorMatchingSnapshot();
  });

  it('should throw an error if an invalid config is present and write is called', async () => {
    const configService = await CheckupConfigService.load(async () => ({
      filepath: '.',
      config: ({
        plugins: {},
        tasks: {
          someTask: 'foo',
        },
      } as unknown) as CheckupConfig,
      format: CheckupConfigFormat.JSON,
    }));

    expect(() => configService.write()).toThrowErrorMatchingSnapshot();
  });

  it('should map a config given a set of mappers', async () => {
    const config = (
      await CheckupConfigService.load(async () => ({
        filepath: '.',
        config: defaultConfig,
        format: CheckupConfigFormat.JSON,
      }))
    )
      .map(
        ...[
          (config: CheckupConfig) => {
            config.plugins.push('plugin-foo');
            return config;
          },
          (config: CheckupConfig) => {
            config.tasks.bar = {
              isEnabled: false,
            };
            return config;
          },
        ]
      )
      .get();

    expect(config.plugins).toStrictEqual(['plugin-foo']);
    expect(config.tasks).toStrictEqual({
      bar: {
        isEnabled: false,
      },
    });
  });

  it.each(Object.keys(CheckupConfigFormat).map(format => [format]))(
    'should write the config applying the given mappers on calling write for %s files',
    async format => {
      const project = new CheckupFixturifyProject('test');
      const filepath = path.join(project.baseDir, '.checkuprc');
      project.writeSync();
      (
        await CheckupConfigService.load(async () => ({
          filepath: path.join(project.baseDir, '.checkuprc'),
          config: defaultConfig,
          format: format as CheckupConfigFormat,
        }))
      )
        .map(
          ...[
            (config: CheckupConfig) => {
              config.plugins.push('plugin-foo');
              return config;
            },
            (config: CheckupConfig) => {
              config.tasks.bar = {
                isEnabled: false,
              };
              return config;
            },
          ]
        )
        .write();

      expect(fs.readFileSync(filepath, 'utf8')).toMatchSnapshot();
      project.dispose();
    }
  );
});
