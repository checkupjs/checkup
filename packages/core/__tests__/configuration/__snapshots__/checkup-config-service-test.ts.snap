// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkup-config-service should throw an error if an invalid config is present and get is called 1`] = `
"Checkup configuration is malformed:
plugins expected type Array<string>, but got {}
tasks.someTask expected type { isEnabled: boolean }, but got \\"foo\\""
`;

exports[`checkup-config-service should throw an error if an invalid config is present and write is called 1`] = `
"Checkup configuration is malformed:
plugins expected type Array<string>, but got {}
tasks.someTask expected type { isEnabled: boolean }, but got \\"foo\\""
`;

exports[`checkup-config-service should write the config applying the given mappers on calling write for JSON files 1`] = `
"{
  \\"plugins\\": [
    \\"plugin-foo\\"
  ],
  \\"tasks\\": {
    \\"bar\\": {
      \\"isEnabled\\": false
    }
  }
}"
`;

exports[`checkup-config-service should write the config applying the given mappers on calling write for JavaScript files 1`] = `
"module.exports = {
  \\"plugins\\": [
    \\"plugin-foo\\"
  ],
  \\"tasks\\": {
    \\"bar\\": {
      \\"isEnabled\\": false
    }
  }
}"
`;

exports[`checkup-config-service should write the config applying the given mappers on calling write for YAML files 1`] = `
"plugins:
  - plugin-foo
tasks:
  bar:
    isEnabled: false
"
`;
